File: src/guide/combat/initiative.py
from __future__ import annotations

import random
from dataclasses import dataclass


@dataclass
class InitiativeEntry:
    name: str
    roll: int
    modifier: int
    total: int


def roll_d20() -> int:
    return random.randint(1, 20)


def roll_initiative(dex_modifier: int) -> InitiativeEntry:
    roll = roll_d20()
    return InitiativeEntry(name="", roll=roll, modifier=dex_modifier, total=roll + dex_modifier)


def sort_initiative(entries: list[InitiativeEntry]) -> list[InitiativeEntry]:
    """Sort descending by total, ties broken by modifier (higher wins)."""
    return sorted(entries, key=lambda e: (e.total, e.modifier), reverse=True)

File: crates/guide-combat/src/initiative.rs
use rand::Rng;

#[derive(Debug, Clone)]
pub struct InitiativeEntry {
    pub name: String,
    pub roll: i32,
    pub modifier: i32,
    pub total: i32,
}

pub fn roll_d20() -> i32 {
    rand::rng().random_range(1..=20)
}

pub fn roll_initiative(dex_modifier: i32) -> InitiativeEntry {
    let roll = roll_d20();
    InitiativeEntry {
        name: String::new(),
        roll,
        modifier: dex_modifier,
        total: roll + dex_modifier,
    }
}

/// Sort entries by total DESC, then modifier DESC, then id string (not applicable here).
pub fn sort_initiative(mut entries: Vec<InitiativeEntry>) -> Vec<InitiativeEntry> {
    entries.sort_by(|a, b| {
        b.total
            .cmp(&a.total)
            .then_with(|| b.modifier.cmp(&a.modifier))
    });
    entries
}
